share_app/src/main/java/org/odk/share/utilities/OdkInstanceUtils.java

```diff
@@ -0,0 +1,56 @@
+package org.odk.share.utilities;
+
+import org.odk.share.application.Share;
+
+import java.io.UnsupportedEncodingException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+
+public class OdkInstanceUtils {
+
+    // single constructed
+    private OdkInstanceUtils() {
+
+    }
+
+    /**
+     * Uniquely identifiable ID for an instance would be a combination of the name of the directory
+     * for the instance and the device id.
+     *
+     * @return String
+     */
+    public static String getInstanceUuid(String instancePath) throws UnsupportedEncodingException, NoSuchAlgorithmException {
+        return encryptSHA1(getDeviceId() + instancePath);
+    }
+
+    private static String convertToHex(byte[] data) {
+        StringBuilder buf = new StringBuilder();
+        for (byte b : data) {
+            int halfbyte = (b >>> 4) & 0x0F;
+            int twoHalves = 0;
+            do {
+                buf.append((0 <= halfbyte) && (halfbyte <= 9) ? (char) ('0' + halfbyte) : (char) ('a' + (halfbyte - 10)));
+                halfbyte = b & 0x0F;
+            } while (twoHalves++ < 1);
+        }
+        return buf.toString();
+    }
+
+    private static String encryptSHA1(String text) throws NoSuchAlgorithmException, UnsupportedEncodingException {
+        MessageDigest md = MessageDigest.getInstance("SHA-1");
+        byte[] textBytes = text.getBytes("iso-8859-1");
+        md.update(textBytes, 0, textBytes.length);
+        byte[] sha1hash = md.digest();
+        return convertToHex(sha1hash);
+    }
+
+    /**
+     * Same logic used within ODK Collect to obtain the device id in a form
+     *
+     * @return String
+     */
+    private static String getDeviceId() {
+        PropertyManager propertyManager = new PropertyManager(Share.getInstance());
+        return propertyManager.getSingularProperty(PropertyManager.DEVICE_ID_PROPERTY);
```

Comment:
Is it okay if we use device-specific information without user consent?

Reply:
@lakshyagupta21 this id never leaves the device as its encrypted. The only access to this method is via a getter `org.odk.share.utilities.OdkInstanceUtils#getInstanceUuid` which is a SHA1 encrypted algorithm

Reply:
@lakshyagupta21 the reason we used a hash of the device ID was to guarantee that there wouldn't be collisions in the directory names for the submissions (in the receiving device) and at the same time we're able to always recreate the same directory name for a submission. We could, however, switch up the logic; maybe generate UUIDs for the submissions. We'll have to save these in share's SQLite database though.

Reply:
Could we use the UUID already in the submission.xml? Or will that introduce more overhead by having to read the submission XML to get the UUID for each submission?

Reply:
@ukanga The new UUID we are proposing will be the external 'Device ID'  instead of using the actual device id. Its to be used by the receiving device to track all submissions by the same device to control duplication.
The submission UUID is generated with every new submission so its impossible to use it to track the submissions by the same device. That's why we want to store a new 'Device ID' at the same time addressing @lakshyagupta21 concerns